### ISSUE TITLE
perf: NEAT.Train tape-mode training is too slow — Training_ShouldReduceLoss times out at 120s

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1392]
## Summary

\`NEATTests.Training_ShouldReduceLoss\` (the model-family invariant that trains a NEAT — NeuroEvolution of Augmenting Topologies — model for the base 100 \`MemorizationTaskIterations\` count and asserts loss decreased by ≥ 1%) consistently times out at the \`[Fact(Timeout=120000)]\` budget. Surfaced while working [#1304 cluster-6](https://github.com/ooples/AiDotNet/issues/1304); shape-of-the-gap is similar to [#1390](https://github.com/ooples/AiDotNet/issues/1390) (DCGAN GAN training perf gap).

## Repro

\`\`\`
$ dotnet test --filter \"FullyQualifiedName=AiDotNet.Tests.ModelFamilyTests.NeuralNetworks.NEATTests.Training_ShouldReduceLoss\" --framework net10.0
Failed!  - Failed: 1, Passed: 0
  Test execution timed out after 120000 milliseconds
\`\`\`

## Why this is perf, not correctness

NEAT trains via evolutionary topology mutation + per-genome backprop steps. The base-class \`Training_ShouldReduceLoss\` invariant runs 100 \`network.Train(input, target)\` calls. On real PyTorch-style implementations (where the topology is fixed and only the weights evolve) this completes in seconds. Here it doesn't finish in 2 minutes, which means \~1 s+ per Train call on a tiny tabular input. The model's per-genome population evaluation + speciation pass dominates, not the per-genome weight update.

This is the same class of perf gap [#1390](https://github.com/ooples/AiDotNet/issues/1390) calls out for DCGAN tape-mode training: the model is functionally correct but the throughput is ~5-50× off PyTorch CPU baseline at this scale. Closing the gap requires Tensors-side Conv2D/MatMul backward perf work + per-op tape-recording allocator pressure reduction, neither of which is in cluster-6's scope.

## Affected tests (current)

- \`NEATTests.Training_ShouldReduceLoss\`

Adjacent tests at risk if NEAT's per-step cost is unchanged:
- \`NEATTests.MoreData_ShouldNotDegrade\` (250-iter chain)
- \`NEATTests.LossStrictlyDecreasesOnMemorizationTask\` (100-iter chain)
- Any other model-family invariant that exceeds the 60-120 s envelope at \~1 s/Train.

## NOT this issue

- The Occupancy memorization invariant is fixed by [PR #1391](https://github.com/ooples/AiDotNet/pull/1391) (Dropout removal from default layer stack) — different cluster-6 root cause.
- The DenseNet \`MoreData_ShouldNotDegrade\` Adam-overshoot divergence is its own follow-up — not a perf gap, an optimizer-default issue.

## Expected scope

Multi-week perf project. Phase 1: confirm via substep profile (per the [#1390](https://github.com/ooples/AiDotNet/issues/1390) template) which subsystem dominates in NEAT.Train (population eval / speciation / per-genome backprop). Phase 2 + 3: depend on Phase 1.

**Comment by ooples** [comment]
Opened #1419 with a first-pass perf win:

**Local timing on Training_ShouldReduceLoss (single-test, 32-core host):**
- Baseline: 54.0 s  
- Post-fix: 46.0 s (~15% faster, ~74 s under the 120 s budget)

Two issues fixed:
1. The fitness function ran `_population.OrderByDescending(g => g.Fitness).FirstOrDefault()` on every genome eval — O(N) sort × 150 × 50 × 30 = ~34M ops per test. The branch was also semantically broken (probed pre-evaluation best) and dead code per the post-evolution recompute comment immediately below it. Deleted.
2. `ActivateGenome` rebuilt `SortConnectionsTopologically` (O(E²)) from scratch every call. Now cached on the genome keyed on `(Connections.Count, ulong bitmask of IsEnabled)` — weight mutations leave it valid, topology mutations invalidate.

Leaving the issue open — the remaining gap is the multi-week work the issue itself calls out (Dictionary<int,T> allocator pressure in ActivateGenome, per-mutation Clone path, speciation scans). This PR just closes the easy win.

**Comment by ooples** [comment]
Fixed by PR #1419 (`perf(#1392): remove O(N²) ordering in NEAT fitness + cache topology sort`), merged to master.

Verified on latest master: `NEAT.cs` now caches the topologically-sorted node order with a `Connections.Count`-keyed invalidation (see the `Issue #1392 perf helper` blocks at lines ~1013–1129). The per-evaluation O(E²)/O(N²) re-sort is gone — repeated `Predict`/fitness calls hit the cache. Closing as completed (merge didn't auto-close).
